### PR TITLE
feat(Sonic Retro): add presence

### LINKED ISSUE #5889: Sonic Retro | *sonicretro.org
### Website name

Sonic Retro

### Website URL

https://sonicretro.org (main page/news page), https://info.sonicretro.org (wiki page), https://forums.sonicretro.org (forum page)

### Prerequisites

- [ ] It is a paid service
- [ ] It displays NSFW content
- [ ] It is region restricted

### Description

Well, on the main page it could show the news title and the time you have been on the page. On the wiki page it would look like sites like Wikipedia and Fandom, and on the forum page it could show the post and the author's name.

### PR DESCRIPTION
**Pull request by BigBoi-TM**
**Describe the changes in this pull request:**
#
# (Feature) New presence for Sonic Retro
Resolves #5889
Features:
- Detects main pages of each subdomain `sonicretro.org`, `info.sonicretro.org`, & `forums.sonicretro.org`
    - `smallImageKey` displays an icon for each subdomain
- `sonicretro.org`:
    -  Detects home pages, search pages (by category, date, or custom keywords), and articles
    - A button appears when user is on an article (links to the article being read)
- `info.sonicretro.org`:
    - Detects home pages, search pages (by custom keywords), and articles
- `forums.sonicretro.org`:
    - Detects home pages, search pages (by custom keywords), and forums, and threads
#
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

<details>
<summary>Screenshots</summary>

## - Detection of Homepages
<details>
<summary>Pics</summary>

`v sonicretro.org v`
image
`v info.sonicretro.org v`
image
`v forums.sonicretro.org v`
image
</details>

## - `sonicretro.org`
<details>
<summary>Pics</summary>

#### Homepage
image

#### Search By Month
image

#### Search By Date
image

#### Article
image

</details>

## - `info.sonicretro.org`
<details>
<summary>Pics</summary>

#### Homepage
image

#### Search
image
image

#### Special Pages
image

#### Article
image

</details>

## - `forums.sonicretro.org`
<details>
<summary>Pics</summary>

#### Homepage
image

#### Search
image
image

#### Forum
image

#### Thread
image

#### Member Page
image

</details>
</details>

### CODE REVIEW

File: websites/S/Sonic Retro/dist/metadata.json
Comment (ImRodry) [inline websites/S/Sonic Retro/dist/metadata.json:9]: I believe these are unnecessary
```suggestion
		"en": "Second only to Sega - Sonic Retro is the Sonic scene's GFDL input and output wiki resource for collecting data about Sonic the Hedgehog that is open to the entire community."
```

File: websites/S/Sonic Retro/dist/metadata.json
Comment (ImRodry) [inline websites/S/Sonic Retro/dist/metadata.json:12]: you only support info.* and forums.* so you should add these to the URL array instead of making a regexp

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:4]: you're flooring an integer, you probably wanna do this instead
```suggestion
	browsingTimestamp = Math.floor(Date.now() / 1000);
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:13]: This isn't needed

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:17]: did you come across a situation where the search query was empty? How did that happen?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:22]:
```suggestion
						document.querySelector("#main > div.archive-box > h1").textContent
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:26]: this is not valid JS

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:35]: Please avoid using the document title
```suggestion
				presenceData.details = `Looking through ${document.querySelector("#main > div.archive-box > h1").textContent}`;
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:39]: you probably wanna match the URL instead, assuming you're talking about pages of this nature: "https://sonicretro.org/2022/05/02/". You can use something like `/\d{4}\/\d{2}\/\d{2}/.test(document.location.pathname)`

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:42]: why in/on?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:51]: never use String.match for if cases
```suggestion
			} else if (
				/\/([0-9]+(\/[0-9]+)+)\/+((\w|-|\d)*)\//g.test(location.href)
			) {
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:96]: this is invalid JS

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:153]: this is invalid JS

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:267]: all of these conditions except for the second to last one are doing the exact same, you can probably improve this

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:17]: > did you come across a situation where the search query was empty? How did that happen?

Website got a weird bug where u can just press enter when focused on the search bar, that's when the search comes back empty. Just there in case it happens.

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:4]: > you're flooring an integer, you probably wanna do this instead
> 
> ```suggestion
> 
> 	browsingTimestamp = Math.floor(Date.now() / 1000);
> 
> ```

My bad, used an old presence as a template and forgot to change it 😅

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:13]: > This isn't needed

Which part, the image or the URL?

Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:13]: the URL on the small image text

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:17]: in that case they're not searching for anything

Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:17]: > in that case they're not searching for anything

It still triggers the search page, so just to save the code from errors ig. Can it stay?

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:13]: > the URL on the small image text

Should the others go too?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:17]: Sure, but use URLSearchParams instead because there may be other query options

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:38]: Can I see an example of this with the variable replaced?

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:35]: > you probably wanna match the URL instead, assuming you're talking about pages of this nature: "https://sonicretro.org/2022/05/02/". You can use something like `/\d{4}\/\d{2}\/\d{2}/.test(document.location.pathname)`

But u can also sort posts by just month, would I have to make another case?

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:38]: > Can I see an example of this with the variable replaced?

image
image

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:35]: you can make the last part of the regex optional

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:38]:
```suggestion
				presenceData.details = `Looking for posts from ${
```

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:24]: > this is not valid JS

Well then how do I go through with this line without making another `else if` statement that does the same thing; how would I rewrite this?

Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:24]: you should use a logical or properly

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:147]: What I don't get is that not only does this work, prettier even asks to put it back once it's gone. 
image
So where do I go from here?

**Comment by BigBoi-TM** [comment]
Just waiting for a response to a question I had, so plz don't take this as inactivity; just waiting for directions.

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:13]: > I don’t understand what you mean

The other `smallImageText`'s for the other cases

Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:13]: 👀

Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:13]: this wasn't addressed. Yes you should remove all smallImageTexts that contain a URL

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:17]: this wasn't addressed

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:42]: you can just remove this

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:82]: this is not valid, just remove the or

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:79]: remove

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:86]: this will never be true iirc

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:92]: just use .includes()

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:151]: this line is useless

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:168]: you can combine these two with the one above where you also say its unsupported

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:189]: you sure this works? Can you show me a URL with this?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:194]:
```suggestion
					presenceData.details = "Browsing watched threads";
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:195]: find**e**?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:197]: sure?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:200]: nice spellage
```suggestion
					presenceData.details = `Browsing through ${forumTitle.toLowerCase()}`;
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:201]: why the quotes

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:211]: why

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:235]: this is invalid for the 10000th time

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:256]: combine these

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:13]: All right that was all I was asking, but u never responded so I resolved to get ur attention on the matter. Will fix when I get to a computer.

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:17]: > this wasn't addressed

There's nothing to address here, there's NEVER any instance of other query options on this page.

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:86]: Incorrect:

image

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:211]: I used to get errors when I didn't put a string and just had a variable. This presence has been sitting for a min so I haven't gone back and changed these.

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:189]: I misspelled forums 😤

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:197]: These were rushed last min changes 😂😭

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:16]: That can change at any time and the user can also add those yourself. You are **not** allowed to resolve conversations without addressing them for the record

Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts]: changed it

**Comment by BigBoi-TM** [comment]
> You don't need to update your branch this frequently lol

Commit my stuff from vsc, so it sorta does that with each commit I do if they out of sync 😅

**Comment by ImRodry** [comment]
It really doesn’t do that

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:15]: Also just in case this is used elsewhere in the future I'd recommend declaring the search params variable outside the switch statement along with the others
Also not sure if the parenthesis are necessary, remove them if prettier complains
```suggestion
				if (!(new URLSearchParams(search).get("s")))
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:24]:
```suggestion
				(!search && (pathname === "/" || pathname === "/#" || pathname.includes("/page/"))
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:44]: Declare href in the object destructure above
```suggestion
					href
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:35]: now this just isn't smart
```suggestion
					pathname
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:260]:
```suggestion
						search
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:262]:
```suggestion
					href ===
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:263]: I **highly** discourage you to compare against a URL this big, you should either check sth in the page or parts of the URL because this is very volatile

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:83]: please don't compare against hrefs

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:131]:
```suggestion
				presenceData.state =
					document.title.split("Sonic Retro:")[1].split(" - Sonic Retro")[0];
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:125]:
```suggestion
				presenceData.state =
					document.title.split("Talk:")[1].split(" - Sonic Retro")[0];
```

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:151]: invalid JS

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:192]: This is repeated 13 time along this presence, surely you can spare some of those

**Review by ImRodry — CHANGES_REQUESTED** [review]
Please avoid repeating parts of your code as much as possible

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:35]: I dont even remember putting that, my bad boss

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:191]: Figuring an approach to this review; please bear with me.

**Comment by BigBoi-TM** [comment]
> It really doesn’t do that

What shows up before it sends both actual commit and merge commit:
image
The result:
image
So yes Rodry, it does in fact _do that_

**Comment by ImRodry** [comment]
That means you need to pull before you push your commits, you don't have to update your branch every time. Syncing in VSCode simply means pushing or pulling. Also when you have commits to push and you forgot to pull you should do `git pull --rebase` in the command line first

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [inline websites/S/Sonic Retro/presence.ts:81]: did you really mean to put the !search check here or should it go on the entire if statement?

File: websites/S/Sonic Retro/presence.ts
Comment (ImRodry) [reply websites/S/Sonic Retro/presence.ts:191]: just set it outside the if and inside this case

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:81]: No, just to that part.

File: websites/S/Sonic Retro/presence.ts
Comment (BigBoi-TM) [reply websites/S/Sonic Retro/presence.ts:198]: Nah, what I had was right

**Comment by ImRodry** [comment]
Rebased to apply the latest changes in the prettier configuration. I also squashed all commits in your branch as the commit history was quite messy. If you need help with your local branch, please ping me on the #presence-dev channel on Discord.

**Review by ImRodry — APPROVED** [review]
Could be better could be worse, LGTM

**Review by EncryptedDev — APPROVED** [review]

**Comment by EncryptedDev** [comment]
Cheers for the PR

**Pull request merged**
